src/System.Runtime.Analyzers/Core/AvoidZeroLengthArrayAllocations.cs

```diff
@@ -1,39 +1,88 @@
-                                                                             description: s_localizableDescription,
-                                                                             helpLinkUri: null,     // TODO: add MSDN url
-                                                                             customTags: WellKnownDiagnosticTags.Telemetry);
-
-        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
-
-        public override void Initialize(AnalysisContext analysisContext)
-        { 
-            
+
+        /// <summary>The diagnostic descriptor used when Array.Empty should be used instead of a new array allocation.</summary>
+        internal static readonly DiagnosticDescriptor UseArrayEmptyDescriptor = new DiagnosticDescriptor(
+            RuleId,
+            s_localizableTitle,
+            s_localizableMessage,
+            DiagnosticCategory.Performance,
+            DiagnosticSeverity.Warning,
+            isEnabledByDefault: true);
+
+        /// <summary>Gets the set of supported diagnostic descriptors from this analyzer.</summary>
+        public sealed override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics
+        {
+            get { return ImmutableArray.Create(UseArrayEmptyDescriptor); }
+        }
+
+        public sealed override void Initialize(AnalysisContext context)
+        {
+            // When compilation begins, check whether Array.Empty<T> is available.
+            // Only if it is, register the syntax node action provided by the derived implementations.
+            context.RegisterCompilationStartAction(ctx =>
+            {
+                INamedTypeSymbol typeSymbol = ctx.Compilation.GetTypeByMetadataName(ArrayTypeName);
+                if (typeSymbol != null && typeSymbol.DeclaredAccessibility == Accessibility.Public)
+                {
+                    IMethodSymbol methodSymbol = typeSymbol.GetMembers(ArrayEmptyMethodName).FirstOrDefault() as IMethodSymbol;
+                    if (methodSymbol != null && methodSymbol.DeclaredAccessibility == Accessibility.Public &&
+                        methodSymbol.IsStatic && methodSymbol.Arity == 1 && methodSymbol.Parameters.Length == 0)
+                    {
+                        ctx.RegisterOperationAction(AnalyzeOperation, OperationKind.ArrayCreationExpression);
+                    }
+                }
+            });
         }
+
+        private void AnalyzeOperation(OperationAnalysisContext context)
+        {
+            IArrayCreationExpression arrayCreationExpression = context.Operation as IArrayCreationExpression;
+
+            // We can't replace array allocations in attributes, as they're persisted to metadata
+            // TODO: Once we have operation walkers, we can replace this syntactic check with an operation-based check.
+            if (arrayCreationExpression.Syntax.Ancestors().Any(IsAttributeSyntax))
```

Comment:
Not sure what that check would be. I could register an operationblockstart and then the operation action but that would also run over attribute statements I think.